### PR TITLE
Add destination aware log level filtering

### PR DESCRIPTION
**Pull request by KrystalDelusion**

### CODE REVIEW

File: src/yosys_mau/task_loop/logging.py [hunk 1]
```diff
@@ -117,16 +144,28 @@ class LogContext:
...
+    # context manager doesn't like properties that get passed by reference?
+    __levels: LevelDict | None = None
 
-    This does not stop `LogEvent` of smaller levels to be emitted. It is only used to filter which
-    messages to actually print/log. Hence, it does not affect any user installed `LogEvent`
-    handlers.
+    @property
+    def levels(self) -> LevelDict:
+        if not self.__levels:
+            self.__levels = LevelDict()
```
Comment (KrystalDelusion) [inline src/yosys_mau/task_loop/logging.py:153]: I tried to have `levels = LevelDict()` but then the value was persisting between tests.  This is a work around to avoid that, but there's probably something in the context manager that should be doing deep copies instead?

File: src/yosys_mau/task_loop/logging.py [hunk 2]
```diff
@@ -117,16 +144,28 @@ class LogContext:
...
+        return self.__levels
 
-    When logging to multiple destinations, currently there is no way to specify this per
-    destination.
-    """
+    @property
+    def level(self) -> Level:
+        """The minimum log level to display/log.
```
Comment (KrystalDelusion) [inline src/yosys_mau/task_loop/logging.py:158]: These doc strings aren't showing up on sphinx, I need to look into that.

Comment (KrystalDelusion) [reply src/yosys_mau/task_loop/logging.py:158]: Without the `@task_context` decorator they work fine...

File: src/yosys_mau/task_loop/context.py [hunk 3]
```diff
@@ -205,6 +205,7 @@ def __getitem__(self, key: K) -> V:
             else:
                 if isinstance(value, _MISSING_TYPE):
                     raise KeyError(repr(key))
+                return value
```
Comment (jix) [inline src/yosys_mau/task_loop/context.py:208]: Nit, optional: I'd prefer fixes for functionality not added in a PR to be a separate commit as that makes things like bisection and cherry picking easier. (I don't think I'm very good at consistently following my own advice regarding this, though, and especially with mau where there's not too much concurrently ongoing development it's not really an issue.)

File: src/yosys_mau/task_loop/logging.py [hunk 4]
```diff
@@ -117,17 +120,17 @@ class LogContext:
     When set, the downgrading happens before the `LogEvent` is emitted.
     """
 
-    level: Level = "info"
-    """The minimum log level to display/log.
+    dest_levels: TaskContextDict[str, Level] = TaskContextDict()
+    """The minimum log level to display/log for named destinations.
```
Comment (jix) [inline src/yosys_mau/task_loop/logging.py]: It's `dest_levels` here but `destination_label` below. It should be either consistently abbreviated or not abbreviated at all. Either choice works for me, but I'd probably use `destination`.

File: src/yosys_mau/task_loop/logging.py [hunk 5]
```diff
@@ -296,7 +299,10 @@ def log_exception(exception: BaseException, raise_error: bool = True) -> LoggedE
 
 
 def start_logging(
-    file: IO[Any] | None = None, err: bool = False, color: bool | None = None
+    file: IO[Any] | None = None,
+    err: bool = False,
+    color: bool | None = None,
+    destination_label: str | None = None,
```
Comment (jix) [inline src/yosys_mau/task_loop/logging.py:311]: Inconsistent `dest_` vs `destination_` see previous comment.

File: tests/task_loop/test_logging.py [hunk 6]
```diff
@@ -173,7 +173,7 @@ def main():
         tl.log_debug("line 3")
 
         del tl.LogContext.dest_levels["varied"]
-        tl.LogContext.dest_levels[None] = "warning"
+        tl.LogContext.dest_levels[None] = "warning"  # type:ignore
```
Comment (jix) [inline tests/task_loop/test_logging.py:176]: Why is `# type:ignore` needed here? I think `None` actually wouldn't be allowed here given the type signature for `dest_levels` so is there a reason to test it?

File: src/yosys_mau/task_loop/logging.py [hunk 7]
```diff
@@ -147,6 +145,14 @@ class LogContext:
     Like `log_format` this is looked up by the log writing task, not the emitting task.
     """
 
+    dest_levels: TaskContextDict[str, Level] = TaskContextDict()
```
Comment (jix) [inline src/yosys_mau/task_loop/logging.py:148]: This commit moved this line, see the comment on the commit that added this.

**Review by jix — CHANGES_REQUESTED** [review]
Made a quick review pass over the PR and left inline comments. With `dest` vs `destination` resolved this LGTM . Haven't done any extra testing, but CI with the added tests should be sufficient.

File: src/yosys_mau/task_loop/context.py (hunk 3 above)
Comment (KrystalDelusion) [reply src/yosys_mau/task_loop/context.py:208]: I had to add this to prevent the log handler from hanging when it tried to access the default level, but that may not be necessary with the switch to task context dict.  I'll try remove this and see if it works.

Comment (jix) [reply src/yosys_mau/task_loop/context.py:208]: From the surrounding code it looks to me like your change fixes an actual bug and I wasn't suggesting to remove the fix from the PR but rather to have a separate commit containing just the fix, still included in this PR, so that the commit log makes it clear that a bug was fixed.

Although if you're less certain about whether it's an actual bug it's probably a good idea for me to also double check how this was intended to interact with the overall implementation. And maybe also add a regression test that covers that specific case, as, no matter which variant is correct, it's slightly concerning that all existing tests seem pass with both variants.

File: tests/task_loop/test_logging.py (hunk 6 above)
Comment (KrystalDelusion) [reply tests/task_loop/test_logging.py:176]: Accidentally leftover from when I was using `None` as the default value

**Comment by KrystalDelusion** [comment]
Test gets stuck in an infinite loop (pending #17 update)

**Comment by KrystalDelusion** [comment]
That should work now, I reverted the context dict fix on this branch instead of rebasing it out, so I had to drop that commit when rebasing onto main.

**Review by jix — APPROVED** [review]
